sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVOptions.scala

```diff
@@ -183,7 +180,9 @@ class CSVOptions(
       Some(parameters.getOrElse("timestampFormat",
         s"${DateFormatter.defaultPattern}'T'HH:mm:ss.SSSXXX"))
     } else {
-      parameters.get("timestampFormat")
+      // Use Iso8601TimestampFormatter (with strict timestamp parsing) to
+      // avoid parsing dates in timestamp columns as timestamp type
```

Comment:
why do we need this change? when inferring schema, we always try to infer as date then timestamp, right?

Reply:
- There are two types of timestampFormatter: Default and Iso8601.
  - The default formatter use same parsing logic as CAST, which can parse a date value as timestamp.
  - The Iso8601 has more restrict parse. 
  - If not timestampFormat is given, we will use the default one 
  
- In the case of a column containing timestamp value first and dates following, if we use the default formatter, it would still parse the dates as timestamp type, which is not consistent with what we are proposing now.

Reply:
I see, but the CAST code is supposed to be more efficient than the formatter. Do we have some perf numbers? Is there a slowdown for schema inference with timestamp value only CSV files?

Reply:
Let me think about it as well as there are different fallbacks with the regard to date/timestamp parsing. IMHO, we would want to have some kind of flag to revert to the original behaviour but this would be a lot of flags to configure. 

It is possible that some users might be relying on a more lenient parsing of timestamps so if we switch to ISO8601 only, some of the jobs might need to be updated. Let me think a bit more on this.

Reply:
Totally agree with your concerns @cloud-fan & @sadikovi.

After some quick discussion within my team, we agreed on not changing these lines to avoid unnecessary regressions and any other behavior changes. Thus, the behavior after this PR become:
- If user provides a `timestampFormat/timestampNTZFormat`, we will strictly parse fields as timestamp according to the format. Thus, columns with mixing dates and timestamps will be inferred as `StringType`.
- If no `timestampFormat/timestampNTZFormat` specified by user, columns with mixing dates and timestamps could be inferred as `TimestampType/TimestampNTZType`, which is okay and even a good feature.

Does this make sense to you? @sadikovi @cloud-fan

Reply:
Made some follow-up changes, please check the updated description for the behavior after changes and semantics.